### DISCUSSION TITLE
Recommanded configuration for LiveViews testing seems useless

### DISCUSSION THREAD

**Discussion by pierrelegall** [keathley/wallaby#757] [General]
The recommended configuration below (found in the README) is for LiveViews testing. However, it seems useless to me: my tests still works (for now) without it. Note I use `phoenix_live_view` `v0.20.3`. What's its point?

> # LiveView
> 
> In order to test Phoenix LiveView (as of [version 0.17.7](https://github.com/phoenixframework/phoenix_live_view/releases/tag/v0.17.7)) with Wallaby you'll also need to add a function to each mount/3 function in your LiveViews, or use the on_mount live_session lifecycle hook in the router:
> 
> ```elixir
> defmodule MyApp.Hooks.AllowEctoSandbox do
>   import Phoenix.LiveView
>   import Phoenix.Component
> 
>   def on_mount(:default, _params, _session, socket) do
>     allow_ecto_sandbox(socket)
>     {:cont, socket}
>   end
> 
>   defp allow_ecto_sandbox(socket) do
>     %{assigns: %{phoenix_ecto_sandbox: metadata}} =
>       assign_new(socket, :phoenix_ecto_sandbox, fn ->
>         if connected?(socket), do: get_connect_info(socket, :user_agent)
>       end)
> 
>     Phoenix.Ecto.SQL.Sandbox.allow(metadata, Application.get_env(:your_app, :sandbox))
>   end
> end
> ```
> 
> and then including the function usage in the router:
> 
> ```elixir
> live_session :default, on_mount: MyApp.Hooks.AllowEctoSandbox do
>   # ...
> end
> ```

**Comment by mhanberg** [comment]
Are you running your tests async: true and with ecto?

You need this in order for your live view process to connect to the same db connection as your test, which is what allows transactional tests.

**Reply by pierrelegall** [reply]
Ohw I see! It is because my Wallaby tests were configured with `async: false` :wink:

What do you think of a change like this in the README?:

> # LiveView
> 
> In order to run tests on a Phoenix LiveViews (with `{:phoenix_live_view, ">= 0.17"}`) asynchronously with Wallaby, you'll need to add a hook like this one: 
> 
> ```elixir
> defmodule MyApp.Hooks.AllowEctoSandbox do
>   import Phoenix.LiveView
>   import Phoenix.Component
> 
>   def on_mount(:default, _params, _session, socket) do
>     allow_ecto_sandbox(socket)
>
>     {:cont, socket}
>   end
> 
>   defp allow_ecto_sandbox(socket) do
>     %{assigns: %{phoenix_ecto_sandbox: metadata}} =
>       assign_new(socket, :phoenix_ecto_sandbox, fn ->
>         if connected?(socket), do: get_connect_info(socket, :user_agent)
>       end)
> 
>     Phoenix.Ecto.SQL.Sandbox.allow(metadata, Application.get_env(:your_app, :sandbox))
>   end
> end
> ```
> 
> on your LiveView `live_session` `on_mount`:
> 
> ```elixir
> live_session :default, on_mount: MyApp.Hooks.AllowEctoSandbox do
>   # ...
> end
> ```
> 
> **or** on the `on_mount` on the LiveView:
> 
> ```elixir
> on_mount MyApp.Hooks.AllowEctoSandbox
> ```

**Reply by mhanberg** [reply]
it's already under the section describing concurrent testing with Phoenix. I'm not sure it really would make much of a difference.

But, The README is a little long in general, so I think converting it into better organized Guides in hexdocs would be a more helpful change.

**Reply by pierrelegall** [reply]
The previous "Assets" section does not seems to have something to do with concurrent testing, so it's hard to keep the thread 😅

**Reply by pierrelegall** [reply]
> The README is a little long in general, so I think converting it into better organized Guides in hexdocs would be a more helpful change.

:+1:
